### ISSUE TITLE
AWS::AppConfig::EnvironmentMonitor

### ISSUE THREAD

**Issue by chrisoverzero** [aws-cloudformation/cloudformation-coverage-roadmap#840]
## Scope of Request

AWS::AppConfig::Environment is unwieldy to use with monitored rollback in many configurations, such as alongside AWS Lambda. For the Lambda Function to be able to read the AppConfig Configuration Version, it must be passed (among other things) the AppConfig Environment. To monitor the Function for rollback, the name of the function must be passed to a CloudWatch Alarm, the ARN of which must be passed to the monitor collection in the AppConfig Environment. It is undesirable to explicitly name Lambda Functions, so the circular dependency is difficult to avoid. (The same applies for AWS::Serverless::RestApi, for example.)

## Expected Behavior

It should be possible to break the circular dependency with another resource which attaches the other way, something like:

```yaml
ApiMonitor:
  Type: AWS::AppConfig::EnvironmentMonitor
  Properties:
    ApplicationId: !Ref ConfigApplication
    EnvironmentId: !Ref ConfigEnvironment
    AlarmArn: !GetAtt Api5xxAlarm.Arn
    AlarmRole: !GetAtt ConfigRollbackWatcherRole.Arn
```

## Suggest Specific Test Cases

A template with an AppConfig Application, Environment, and Configuration profile; a Lambda function; and a CloudWatch alarm should display it. Have the environment variables of the Function reference each AppConfig resource, have the alarm reference the Function, and have the AppConfig Environment reference the Alarm.

## Helpful Links to Speed Up Research and Evaluation

It would be great if it worked like Security Group [Egress][] and [Ingress][].

[Egress]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-security-group-egress.html
[Ingress]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-security-group-ingress.html

## Category

Management & Governance

**Comment by dggorsen** [comment]
Thanks for the feature request Chris! Your concerns are certainly valid, and we would like to work with you in order to resolve the circular dependency issues you are experiencing in your template. My initial suggestion would be to leverage resource names in the AppConfig GetConfiguration API rather than resource ids. This would enable you:

1. Set a CloudFormation parameter to be used for your Application/Environment/Configuration names.
2. Create your Lambda while passing it the parameters specified in step 1.
   * Your lambda may then call `get-configuration --application MyApplicationName --environment 'MyEnvironmentName' --configuration 'MyConfigurationName' --clent-confiuration-version '$cached_version' --client-id '$lambda_container_id'`
   * As a side note I would recommend leveraging the [AppConfig Lambda extension](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-integration-lambda-extensions.html) to help you apply best practices in your interactions with AppConfig.
3. Create your alarms that monitor the lambda.
4. Create your AppConfig Application, Environment (with monitors) and ConfigurationProfile by leveraging the name parameters specified in step 1 and avoiding a circular dependency with step 2.

Hope this helps!

Best,
Dillon

**Comment by chrisoverzero** [comment]
Hi, @dggorsen,

I had no idea that names could be used to retrieve AppConfig configurations. It feels workaroundy, but I was able to implement roughly your suggestion. The feature still feels… backwards to me? But it’s good to know that it’s not impossible to make work. 

There continued to be a circular dependency with respect to the IAM Managed Policy, but I flipped it around to a Policy in order to resolve that. We are indeed using the extension. (Now it’s gotta work on arm64! 😛)
